### ISSUE TITLE
Unable to find correct repl with two active repls (Clojure + Clojurescript)

### ISSUE THREAD

**Issue by rpkarlsson** [clojure-emacs/cider#2547]
## Expected behavior
The correct repl to eval the code.

## Actual behavior
No linked repl can be found.

## Steps to reproduce the problem
- Jack in to a Clojure repl
- Connect to a Clojurescript repl
- Jump to the Clojurescript repl
- Jump to a Clojure buffer and eval an expression

### CIDER version information

```
;; CIDER 0.19.0snapshot (package: 20181212.1936), nREPL 0.2.12
;; Clojure 1.9.0, Java 1.8.0_191
```

### Lein/Boot version
Lein 2.8.1

### Emacs version
26.1 

### Operating system
Fedora 29

Seems like `cider-repls` in `cider-connection.el` only considers the last used repl. I have tried to debug but got stuck on the ensure path using `sesman-ensure-session`.
[This fix](https://github.com/rpkarlsson/cider/commit/7726770ca5d75ae2187b78e428d4dd64d3f693c6) will work if `ensure` is not thruthy as it will seq-filter through all active repls not just the last visited.
Still need to find a way to make `sesman-ensure-session` to consider all linked connections. 

By opening this issue I'm hoping to get some help with finding a solution so that evaluating code will consider all my open repls and not just the last one that I visited.

**Comment by rpkarlsson** [comment]
Poking around a bit more it seems like `sesman-ensure-session` gives priority to the latest visited repl. This keeps cider from finding the correct repl when a project has active sessions for both clojure and clojurescript. I've found a work around that seems to be working for my case (linked below). Unsure how this would effect other workflows. Will keep testing and make a PR if you guys are interested.

https://github.com/rpkarlsson/cider/commit/74e6d69c368b3535af89d8dc32f102116f32df48

**Comment by rpkarlsson** [comment]
Ah. Started too look at the testing side of things and noticed some failing test like:
- `cider-repls when multiple sessions exist always returns the most recently used connection`
- `cider-current-repl when multiple sessions exist always returns the most recently used connection`

Seems to be expected to always return the latest repl disregarding if its clj or cljs.
If so, sorry for the issue and taking your time...

**Comment by jherrlin** [comment]
I am facing the same problem.

**Comment by bbatsov** [comment]
I think were supposed to give priority to the most recent session only in case of ambiguity. Perhaps the tests are wrong. The current behaviour seems like a bug to me. 

@vspinu any thoughts on this?

**Comment by vspinu** [comment]
@rpkarlsson could you please provide more info on the reproducible steps. Particularly in which projects do you cljs-connect in and in which project do you select the clj file. 

I do see an issue issue similar to yours with two sessions to the same server (one clj one cljs) but it shows `user-error: No clj REPLs in current session` instead of `no linked sessions` as you described. 

Here is what I do, and it works as expected:

  - Jack in clj in project A (C-c C-x j j)  to create clj-repl-A
  - Run lein repl in project B in terminal
  - Connect cljs to lein-repl from project B (C-c C-x c s) to create cljs-repl-B
  - Go cljs-repl-B
  - Go to some clj file in project A
  - Do eval or C-c C-z and it works as expected in the sense that it evaluates/switches to clj-repl-A

Here is what I do and it throws `No clj REPLs in current session` (in principle also as expected):

  - Jack in clj in project A (C-c C-x j j)  to create clj-repl-A
  - Connect cljs repl from project A (C-c C-x c s) to create cljs-repl-A (you get a warning, press `y` to confirm)
  - Go to cljs-repl-A
  - Go to some clj file in project A
  - Do eval or C-c C-z and it throws `No clj REPLs in current session`

The latter is because your current session is the one which you have created by cljs-connect. Both cljs-repl-A and clj-repl-A sessions are on the same footing and the current one is ambiguous. This issue could be fixed but such connection pattern is discouraged. You really should connect a sibling in order to have one session not two.

But again, none of the above seems to be what you do.

**Comment by rpkarlsson** [comment]
Sorry for the unclear description. 
I'm doing something very similar to your second example. But the problem seems to be with how I use Cider.
I have, since I created the issue,  updated both my dependencies and the way I jack-in/connect to the repls but the use case I'm describing is still the same.

I have one project with both Clj(using Lein) and Cljs(using shadow-cljs). Essentially what I want to do is:
- Jack in clj in project A (Lein same git project) (C-c C-x j j) to create clj-repl-A
- Jack in cljs repl from project A (shadow-cljs same git project) (C-c C-x j s) to create cljs-repl-A (you get a warning, press y)
- Go to cljs-repl-A
- Go to some clj file in project A
- Do C-x C-e on a form

**Expected:**
  Code is evaluated in the clj-repl-A

**Actual:**
  Nothing happens. 
  Doing C-c C-z gives a  `No clj REPLs in current session...` error.

The linked commit changes so that my expected outcome happens. 

But since @vspinu wrote that "This issue could be fixed but such connection pattern is discouraged." It seems like I'm using Cider in the wrong way.
So my conclusion is that I have to either split my clj+cljs project into two projects or live with having to jump to the appropriate repl when switching from clj to cljs or vice versa, this still seems like a problem since as far as I understand still use the discouraged connection pattern.

For the past weeks I've just overridden Cider with the code in the linked commit and this has worked fine for me. But as this connection pattern is discouraged perhaps that's not such a good solution @jherrlin?

Have tried different ways to handle this (siblings, jack-in-clj&cljs) but I'm unable to find anything that works for my use case.

**Comment by vspinu** [comment]
> Have tried different ways to handle this (siblings, jack-in-clj&cljs) but I'm unable to find anything that works for my use case.

When you jack-in twice you create two sessions, each session has one repl. This might be useful in some rare circumstances but mostly it's not what you needed. Normally you would create one session per project with two (or more) REPLs. This is why you get that y-or-n warning in the first place.

Either jack-in-clj&cljs or creating a sibling cljs is the "recommended" way. Why doesn't that work for you? There might be an issue with shadow cljs. There is an issue or two already opened for shadow jack-in problems; this one might be related. 

I can fix the "No clj REPLs ..." error but I am afraid this could encourage the "wrong" pattern and thus obscure issues with the "right" pattern as it seems to be the case here.

**Comment by rpkarlsson** [comment]
I now try:
- C-c C-x j m
- when asked "Which command shoud be used[...]" I choose the default that is `lein`
- I then get prompted for "Select ClojureScript REPL type:" and I choose `shadow`
- The repl starts
- I then get a prompt "Select-cljs build[...]" where I chose the correct build

Two repls are opened. Both in the server namespace (probably using Lein).
One repl throws a "missing instance" exception from `shadow.cljs.devtools.server.runtime.get-instance!`.
I seem to be unable to get the cljs repl connected to shadow.

If I choose `shadow` instead of `lein` in step two above I get two repls in the `shadow.user` ns.
However when I eval code it actually seems to be running a clj-repl and I'm able to eval code in both clj and cljs files! (awesome!)

So it's a user error. 
I was fooled by not being able to choose `lein` + `shadow`.
Another thing that fooled me was the `shadow.user` in both prompts when jacking in with `shadow`.
And lastly as `lein` wasn't run the ns wasn't evaled on start as when connecting with `lein`
This all seems to be working as shadow somehow uses `lein` under the hood so all deps are loaded for both clj and cljs environments.

Below is the results using connect-sibling.

If I connect using `lein` and then C-c C-x s s in a clj-file I get the same error as in steps described above.
If I'm standing in a cljs-file i get `No cljs REPLs in current session`

If I connect using `shadow` and then C-c C-x s s in a cljs-file or a cljs-file i get a clj-repl that works (no ns evaled from lein so need to do that manually)

Previously my workflow was a bit different. 
First I started a lein repl using Cider. Then starting a shadow-cljs repl in a terminal connecting to that using cider connect.

Basically misunderstanding how Cider's supposed to be used.
Thank you @vspinu for clarifying!

**Comment by vspinu** [comment]
I see. It's mostly shadow related then. I am not familiar with shadow but will have a deep dive within a week or two and will be able to better judge what could be done on CIDER side in order to avoid such confusion. Thanks for providing the detailed description of your patterns.

> If I connect using shadow and then C-c C-x s s in a cljs-file or a cljs-file i get a clj-repl that works (no ns evaled from lein so need to do that manually)

Hmm, this one doesn't seem right. If you jacking using shadow `C-c C-x j s` then you should connect a clj sibling with `C-c C-x s j`. The `C-c C-x s s` should make another cljs sibling.

**Comment by rpkarlsson** [comment]
> Hmm, this one doesn't seem right. If you jacking using shadow C-c C-x j s then you should connect a clj sibling with C-c C-x s j. The C-c C-x s s should make another cljs sibling.
A copy-paste error sorry about that. 

Will try actually using Cider the way its intended. Thanks for your time.
